lib/Semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,393 @@
+//===-- lib/semantics/check-select-stmt.cc --------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "check-select-stmt.h"
+#include "flang/Parser/message.h"
+#include "flang/Semantics/tools.h"
+
+namespace Fortran::semantics {
+
+class SelectCaseHelper {
+
+public:
+  SelectCaseHelper(parser::Messages &messages, int kind, TypeCategory category)
+    : kind_{kind}, category_(category), messages_{messages} {}
+
+  template<typename T> std::string GetValueAsString(T val) {
```

Comment:
This function seems like a useful utility; hiding it as a non-static (why?) member function of a helper class in an implementation file will just lead to code duplication later.

A more general function that accepts more types and converts them to expressions and uses AsFortran() would be better.

Reply:
I will try to make it a generic function and move it to formatting.cpp file.

Reply:
Hi klausler, Tim,
Converting the value to a expression and using AsFortran() would print the kind value as well:
select-case.f90:98:7: Previous CASE statement matching .true._1
select-case.f90:118:7: Previous CASE statement matching 81_8:90_8

I plan to use string replace to remove _*. I saw the usage of replace in couple of places.
What is the usual way to avoid kind getting printed along with the value?

Reply:
Here is the current implementation of GetValueAsString and GetValuePairAsString functions:
template<typename T> std::string GetValueAsString(T val) {
  if constexpr (std::is_same_v<T, bool>) {
    return Expr<Type<TypeCategory::Logical, 1>>{
        Constant<Type<TypeCategory::Logical, 1>>{val}}
        .AsFortran();
  } else if constexpr (std::is_same_v<T, std::int64_t>) {
    return Expr<Type<TypeCategory::Integer, 8>>{
        Constant<Type<TypeCategory::Integer, 8>>{val}}
        .AsFortran();
  } else if constexpr (std::is_same_v<T, std::string> ||
      std::is_same_v<T, std::u16string> || std::is_same_v<T, std::u32string>) {
    return parser::QuoteCharacterLiteral(val);
  } else {
    // other unsupported types, Return emtpy string
    return std::string();
  }
}

template<typename T> std::string GetValuePairAsString(T fromVal, T toVal) {
  return ((fromVal ? GetValueAsString(fromVal.value()) : ""s) + ":"s +
      (toVal ? GetValueAsString(toVal.value()) : ""s));
}
Please suggest any better alternative ways if any.

Reply:
> Hi klausler, Tim,
> Converting the value to a expression and using AsFortran() would print the kind value as well:
> select-case.f90:98:7: Previous CASE statement matching .true._1
> select-case.f90:118:7: Previous CASE statement matching 81_8:90_8
> 
> I plan to use string replace to remove _*. I saw the usage of replace in couple of places.
> What is the usual way to avoid kind getting printed along with the value?

I suggest leaving the kind there, at least for now.

Reply:
Thanks for the clarification Tim. I would retain the kind value.